bambi/plots/effects.py

```diff
@@ -0,0 +1,349 @@
+    use_hdi : bool, optional
+        Whether to compute the highest density interval (defaults to True) or the quantiles.
+    hdi_prob : float, optional
+        The probability for the credibility intervals. Must be between 0 and 1. Defaults to 0.94.
+        Changing the global variable ``az.rcParam["stats.hdi_prob"]`` affects this default.
+    transforms : dict, optional
+        Transformations that are applied to each of the variables being plotted. The keys are the
+        name of the variables, and the values are functions to be applied. Defaults to ``None``.
+
+    Returns
+    -------
+    cap_data : pandas.DataFrame
+        A DataFrame with the ``create_cap_data`` and model predictions.
+
+    Raises
+    ------
+    ValueError
+        If passed ``covariates`` is not in correct key, value format.
+        If length of ``covariates`` is not between 1 and 3.
+
+    """
+
+    covariate_kinds = ("horizontal", "color", "panel")
+    if not isinstance(covariates, dict):
+        covariates = listify(covariates)
+        covariates = dict(zip(covariate_kinds, covariates))
+    else:
+        assert covariate_kinds[0] in covariates
+        assert set(covariates).issubset(set(covariate_kinds))
+
+    assert 1 <= len(covariates) <= 3
+
+    if transforms is None:
+        transforms = {}
+
+    if hdi_prob is None:
+        hdi_prob = az.rcParams["stats.hdi_prob"]
+
+    if not 0 < hdi_prob < 1:
+        raise ValueError(f"'hdi_prob' must be greater than 0 and smaller than 1. It is {hdi_prob}.")
+
+    cap_data = create_cap_data(model, covariates)
+    response_name = get_aliased_name(model.response_component.response_term)
+    response_transform = transforms.get(response_name, identity)
+
+    if pps:
+        idata = model.predict(idata, data=cap_data, inplace=False, kind="pps")
+        y_hat = response_transform(idata.posterior_predictive[response_name])
+        y_hat_mean = y_hat.mean(("chain", "draw"))
+    else:
+        idata = model.predict(idata, data=cap_data, inplace=False)
+        y_hat = response_transform(idata.posterior[f"{response_name}_{target}"])
+        y_hat_mean = y_hat.mean(("chain", "draw"))
+
+    if use_hdi and pps:
+        y_hat_bounds = az.hdi(y_hat, hdi_prob)[response_name].T
+    elif use_hdi:
+        y_hat_bounds = az.hdi(y_hat, hdi_prob)[f"{response_name}_{target}"].T
+    else:
+        lower_bound = round((1 - hdi_prob) / 2, 4)
+        upper_bound = 1 - lower_bound
+        y_hat_bounds = y_hat.quantile(q=(lower_bound, upper_bound), dim=("chain", "draw"))
+
+    cap_data["estimate"] = y_hat_mean
+    cap_data["hdi_3%"] = y_hat_bounds[0]
+    cap_data["hdi_97%"] = y_hat_bounds[1]
+
+    return cap_data
+
+
+@dataclass
+class Contrast:
+    term: Union[str, list]
+    value: Union[list, np.ndarray, None]
+
+
+@dataclass
+class Response:
+    name: str
+    target: str
+
+
+@dataclass
+class ContrastEstimate:
+    comparison: Dict[str, xr.DataArray]
+    hdi: Dict[str, xr.Dataset]
```

Comment:
Should these go here? I'm not 100% what are the "best practices" for dataclasses

Reply:
I thought the same thing. I debated putting them either here or in the `utils.py` module. However, these classes aren't used by the other effects functions (predictions or slopes). I decided to use a class instead of a dict to have a strongly typed object with named parameters that have an attached type. Then, I use `@dataclass` to create the constructor with params., and to store the params. into the fields, etc. Aka it lets me to not write as much boilerplate code. 

That was my thought process for using the classes and dataclass decorator.

Reply:
it's ok to leave them here